signimage/signimage.go

```diff
@@ -0,0 +1,357 @@
+	}), nil
+
+}
+
+func main() {
+	// get the env vars we are using for setup, or set some sensible defaults
+
+	var unsignedimagename string
+	var signedimagename string
+	var fileslist string
+	var privkeyfile string
+	var pubkeyfile string
+	var pullsecret string
+	var pushsecret string
+	var nopush bool
+	flag.StringVar(&unsignedimagename, "unsignedimage", "", "name of the image to sign")
+	flag.StringVar(&signedimagename, "signedimage", "", "name of the signed image to produce")
+	flag.StringVar(&fileslist, "filestosign", "", "colon seperated list of kmods to sign")
+	flag.StringVar(&privkeyfile, "key", "", "path to file containing private key for signing")
+	flag.StringVar(&pubkeyfile, "cert", "", "path to file containing public key for signing")
+	flag.StringVar(&pullsecret, "pullsecret", "", "path to file containing credentials for pulling images")
+	flag.StringVar(&pullsecret, "pushsecret", "", "path to file containing credentials for pushing images")
+	flag.BoolVar(&nopush, "no-push", false, "do not push the resulting image")
+
+	flag.Parse()
+
+	checkarg(&unsignedimagename, "UNSIGNEDIMAGE", "")
+	checkarg(&signedimagename, "SIGNEDIMAGE", unsignedimagename+"signed")
+	checkarg(&fileslist, "FILESTOSIGN", "")
+	checkarg(&privkeyfile, "KEYSECRET", "")
+	checkarg(&pubkeyfile, "CERTSECRET", "")
+	checkarg(&pullsecret, "PULLSECRET", "")
+	checkarg(&pushsecret, "PUSHSECRET", pullsecret)
+	// if we've made it this far the arguements are sane
+
+	// get a temp dir to copy kmods into for signing
+	extractiondir, err := os.MkdirTemp("/tmp/", "kmod_signer")
+	if err != nil {
+		fmt.Errorf("could not create temp dir: %v\n", err)
+		panic(err)
+	}
+
+	// sets up a tar archive we will use for a new layer
+	var b bytes.Buffer
+	tarwriter := tar.NewWriter(&b)
+
+	//make a map of the files to sign so we can track what we want to sign
+	kmodstosign := make(map[string]string)
+	for _, x := range strings.Split(fileslist, ":") {
+		kmodstosign[x] = "not found"
+	}
+
+	// set up image download otions
+	opts := make([]name.Option, 0)
+	ref, err := name.ParseReference(unsignedimagename, opts...)
+	if err != nil {
+		fmt.Errorf("could not parse the container image name: %v\n", err)
+		panic(err)
+	}
+
+	a, err := getAuthFromFile(pullsecret, strings.Split(unsignedimagename, "/")[0])
+	if err != nil {
+		fmt.Printf("failed to get auth: %v\n", err)
+		panic(err)
+	}
+
+	descriptor, err := remote.Get(ref, remote.WithAuth(a))
+	if err != nil {
+		te := &transport.Error{}
+
+		if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
+			fmt.Errorf("could not get image: http.StatusNotFound")
+		}
+
+		fmt.Errorf("could not get image: %v", err)
+		panic(err)
+	}
+
+	img, err := descriptor.Image()
+	if err != nil {
+		fmt.Errorf("could not Image(): %v", err)
+		panic(err)
+	}
+	fmt.Printf("\n== Successfully pulled image: %s\n", unsignedimagename)
+	fmt.Printf("\n== Looking for files: %s\n", strings.Replace(fileslist, ":", " ", -1))
```

Comment:
Could we use a logging framework here? [logr](https://github.com/go-logr/logr) as the facade and [klog](https://github.com/kubernetes/klog) as the implementation work well in the community.

Reply:
We could but I don't see what benefit it would add, this is a pretty simple batch job, that runs in its own container, and then exits. It's doing one thing and (hopefully) doing it well.

The complexity of using a full blown logging framework just so I can dump a handful of progress lines to stdout feels excessive to me, great for long running stuff like controllers, overkill for this.

Reply:
The complexity of setting up `klog` together with `logr` is moderate: https://github.com/kubernetes-sigs/kernel-module-management/blob/main/main.go#L78-L82
It would give us fields, levelled logging, the ability to mock the logger in unit tests, and more.
This is also the [preferred approach in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#logging-in-kubernetes):
> Kubernetes project uses [klog](https://github.com/kubernetes/klog) for logging.